Add leader election to controller-manager

shashidharatd opened the pull request:
This PR supersedes https://github.com/kubernetes-sigs/federation-v2/pull/394

- Add leader election logic to controller-manager initialisation.
- ~~Replace statefulset with deployment for controller-manager.~~ This is already handled in https://github.com/kubernetes-sigs/federation-v2/pull/404
- Add an e2e test case in managed setup to test leader election.

/cc @marun @irfanurrehman @fisherxu

marun commented:
@shashidharatd Thank you for doing this!  Will review shortly.  As per comments I made on #404, I think there may need to be 2 modes for federation: leader election (using deployment) and non-leader election (using statefulset).

If #394 is superceded, do you want to close it?

shashidharatd commented:
> As per comments I made on #404, I think there may need to be 2 modes for federation: leader election (using deployment) and non-leader election (using statefulset).

We could handle this in PR #404. What in your opinion, should be the default mode? 
leader election can work with both deployment and statefulset.
while non-leader election mode can only work with statefulset.

marun commented:
@shashidharatd I was under the impression that leader election required a deployment, but if statefulset works too that would be preferable.  Then there could be only one way to deploy a controller manager and leader election could be enabled via a configuration change (and potentially via component config in the future).

marun commented:
@shashidharatd I'm a bit behind in reviews but will make time tomorrow.

yamt commented:
do we intend to support both of deployment and statefulset? why?

irfanurrehman commented on test/e2e/leaderelection.go:
```diff
@@ -0,0 +1,104 @@
+limitations under the License.
+*/
+
+package e2e
```
The test here seems to be more suitable as an in package test. The only dependency that I see used from `e2e framework` is the `config`, which really doesn't warrant usage of `e2e framework`.

marun replied:
This test absolutely belongs in the e2e package.  It is dependent on the managed fixture (test-managed federation composed from kubeapiserver and etcd binaries.

irfanurrehman replied:
Yeah, understood the need for API server now.

irfanurrehman commented on test/e2e/leaderelection.go:
```diff
@@ -0,0 +1,104 @@
+		waitToBecomeLeader(tl, secondaryLeaderElector)
+		tl.Log("Secondary instance is elected leader")
+
+		secondaryContextCancel()
```
[Not necessary for this PR merge, as these scenario might be tested somewhere else, I did not find them in client-go though] I think additional useful tests might be:
1 - Both instances start together, only one of them becomes the leader
2 - Emulate network partitioning, both become leaders, when communication is back, one of them relents leadership

yamt replied:
both become leaders? isn't one of network partitions with etcd quorum expected to win?

irfanurrehman replied:
@yamt, My bad. I did assume this for the controllers running in two different clusters (without looking at the internals of the implementation). You are right that both controllers will need to be synced up to an API server (or multiple API servers) backed by same quorum backed store.

marun replied:
Likely not a solved problem anytime soon for arbitrary environments.  There's a reason raft has a hard upper bound on latency between nodes - that's inherent complexity that will be hard or impossible to solve.

irfanurrehman commented on cmd/controller-manager/app/controller-manager.go:
```diff
@@ -104,6 +107,36 @@ func Run(opts *options.Options) error {
+		}
+	}()
+
+	lec.Run(ctx)
```
Will this loop forever?
A brief look at the package code and its relevant use in k8s seems to indicate that it does not.
I did not find `Run` calling the handler `OnStartedLeading` whenever it gains leadership in a loop.
k8s ctrl-mgr seem to [implement usage](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L252) such that the process exits with error if it loses leadership; the pod should ideally re-start the process and while starting it would wait to gain leadership if there already is an existing leader.
Should we also be doing the same?

marun replied:
Here's the implementation of `Run`:

```golang
func (le *LeaderElector) Run(ctx context.Context) {                                                                          
    defer func() {                                                                                                           
        runtime.HandleCrash()                                                                                                
        le.config.Callbacks.OnStoppedLeading()                                                                               
    }()                                                                                                                      
    if !le.acquire(ctx) {                                                                                                    
        return // ctx signalled done                                                                                         
    }                                                                                                                        
    ctx, cancel := context.WithCancel(ctx)                                                                                   
    defer cancel()                                                                                                           
    go le.config.Callbacks.OnStartedLeading(ctx)                                                                             
    le.renew(ctx)                                                                                                            
}    
```

Check the `acquire` method's use of `JitterUntil`.  Once acquired, `OnStartedLeading` gets called and then `renew` gets called, which will loop until no longer the leader and then exit.  An exit will prompt the container to be restarted, and it will loop in `acquire` until it is elected the master.  I think this looks solid.

irfanurrehman replied:
Yeah, I did miss the simple fact that the container/pod default restart policy is `restartPolicy: Always` and would restart the process on a successful exit  also.

marun commented on cmd/controller-manager/app/leaderelection/leaderelection.go:
```diff
@@ -0,0 +1,83 @@
+	"github.com/kubernetes-sigs/federation-v2/cmd/controller-manager/app/options"
+)
+
+func LeaderElector(opts *options.Options, fnStartControllers func(*options.Options, <-chan struct{})) (*leaderelection.LeaderElector, error) {
```
Consider renaming this factory function to differentiate from `leaderelection.LeaderElector` (e.g. NewFederationLeaderElector).

shashidharatd replied:
Updated

marun commented on cmd/controller-manager/app/leaderelection/leaderelection.go:
```diff
@@ -0,0 +1,83 @@
+)
+
+func LeaderElector(opts *options.Options, fnStartControllers func(*options.Options, <-chan struct{})) (*leaderelection.LeaderElector, error) {
+	restclient.AddUserAgent(opts.Config.KubeConfig, "leader-election")
```
Consider renaming the user agent (e.g. `federation-leader-election`) to ensure the agent is sufficiently differentiated.

shashidharatd replied:
Done

marun commented on cmd/controller-manager/app/leaderelection/leaderelection.go:
```diff
@@ -0,0 +1,83 @@
+	// Prepare event clients.
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: leaderElectionClient.CoreV1().Events(opts.Config.FederationNamespace)})
+	eventRecorder := broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "federation-controller"})
```
'federation-controller-manager'?  Also, consider defining as a constant instead of defining here and on line 56.

shashidharatd replied:
Done

marun commented on cmd/controller-manager/app/options/options.go:
```diff
@@ -50,11 +51,31 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&o.LimitedScope, "limited-scope", false, "Whether the federation namespace will be the only target for federation.")
 	fs.DurationVar(&o.ClusterMonitorPeriod, "cluster-monitor-period", time.Second*40, "How often to monitor the cluster health")
+
+	fs.BoolVar(&o.LeaderElection.LeaderElect, "leader-elect", true, "Start a leader election client and gain leadership before "+
```
(No action required) The addition of so many new configuration parameters suggests to me that implementing component configuration (#114) should be a priority for beta (not for this PR though!).  Thoughts?

shashidharatd replied:
Agree.. we should definitely add it for beta

marun commented on pkg/controller/util/cluster_util.go:
```diff
@@ -45,6 +45,11 @@ const (
 	KubeconfigSecretDataKey = "kubeconfig"
 	getSecretTimeout        = 1 * time.Minute
+
+	LeaseDuration = 15 * time.Second
```
I think more descriptive names are suggested here, e.g. `DefaultLeaderElectionLeaseDuration`.  Reusing the names from `LeaderEelectionConfiguration` doesn't make sense to me because the context of being included in the struct is lost.

shashidharatd replied:
Updated.

marun commented on test/e2e/leaderelection.go:
```diff
@@ -0,0 +1,104 @@
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/test/common"
+	"github.com/kubernetes-sigs/federation-v2/test/e2e/framework"
+	. "github.com/onsi/ginkgo"
```
Please separate the ginkgo import onto its own line.

shashidharatd replied:
Fixed.

marun commented:
@shashidharatd If you aren't intending to add documentation for this feature in this PR, can you please ensure that the issue mentions it and remains open until the docs merge?

shashidharatd commented on cmd/controller-manager/app/controller-manager.go:
```diff
@@ -104,6 +107,36 @@ func Run(opts *options.Options) error {
+		return errors.New("finished without leader election")
+	}
+
+	lec, err := leaderelection.LeaderElector(opts, startControllers)
```
Done

shashidharatd commented:
> @shashidharatd I was under the impression that leader election required a deployment, but if statefulset works too that would be preferable. Then there could be only one way to deploy a controller manager and leader election could be enabled via a configuration change (and potentially via component config in the future).

@marun, Let us discuss this in PR #404 

> do we intend to support both of deployment and statefulset? why?

@yamt, i would prefer to use any one method which makes sense.

> @shashidharatd If you aren't intending to add documentation for this feature in this PR, can you please ensure that the issue mentions it and remains open until the docs merge?

Added comment in https://github.com/kubernetes-sigs/federation-v2/issues/402#issuecomment-471397406

@irfanurrehman, hope your comments are addressed. @marun, have addressed your comment. 
PTAL once again. Thanks!

irfanurrehman commented:
> @irfanurrehman, hope your comments are addressed. @marun, have addressed your comment.
> PTAL once again. Thanks!

yes, looks good to me.. good work indeed @shashidharatd!

marun commented on cmd/controller-manager/app/leaderelection/leaderelection.go:
```diff
@@ -0,0 +1,84 @@
+	"github.com/kubernetes-sigs/federation-v2/cmd/controller-manager/app/options"
+)
+
+func NewFederationLeaderElector(opts *options.Options, fnStartControllers func(*options.Options, <-chan struct{})) (*leaderelection.LeaderElector, error) {
```
(No action required) Please avoid using type prefixes.  I don't think they add value when used in the name of a typed variable.  A more common golang idiom would be <name>Func.

marun commented on cmd/controller-manager/app/options/options.go:
```diff
@@ -50,11 +51,31 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
+
+	fs.BoolVar(&o.LeaderElection.LeaderElect, "leader-elect", true, "Start a leader election client and gain leadership before "+
+		"executing the main loop. Enable this when running replicated controller-manager for high availability.")
+	fs.DurationVar(&o.LeaderElection.LeaseDuration, "leader-elect-lease-duration", util.DefaultLeaderElectionLeaseDuration, ""+
```
(No action required) I'm assuming you copied this from somewhere, but I'd rather not use string concatenation to define multiline strings if at all possible.  Editing them is not fun.

marun commented on test/e2e/leaderelection.go:
```diff
@@ -0,0 +1,108 @@
+
+func waitToBecomeLeader(tl common.TestLogger, lec *leaderelection.LeaderElector) {
+	err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+		if lec.IsLeader() {
```
(No action required)

```return lec.IsLeader(), nil```

marun commented on test/e2e/leaderelection.go:
```diff
@@ -0,0 +1,108 @@
+})
+
+func waitToBecomeLeader(tl common.TestLogger, lec *leaderelection.LeaderElector) {
+	err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
```
Is there a reason these timings should differ from the defaults used by all the other tests?

shashidharatd replied:
forgot to replace in the final version. was doing so in local testing. should have been little careful. Thanks for your meticulous review.

marun commented:
/lgtm
